### PR TITLE
feat: Allow labels to be specified on `dagger session`. Add default labels.

### PR DESCRIPTION
**Pull request by grouville**
Context present in top-level issue [Append SDK info when downloading the Engine image](https://github.com/dagger/dagger.io/issues/1985)

### Problem
Before this PR, the `dagger session` subcommand did not have a way to forward information from the SDKs to the engine:
```shell
$ ./dagger session --help
WARNING: this is an internal-only command used by Dagger SDKs to communicate with the Dagger engine. It is not intended to be used by humans directly.

Usage:
  dagger session [flags]

Flags:
  -h, --help   help for session

Global Flags:
      --debug            show buildkit debug logs
      --workdir string   The host workdir loaded into dagger (default ".")
```

### Solution
This PR adds a hidden variadic `--label` flag implementing this possibility.
It also constructs the final user-agent to send to the registry and appends it to the query to the registry.

```shell
./dagger session --help
WARNING: this is an internal-only command used by Dagger SDKs to communicate with the Dagger Engine. It is not intended to be used by humans directly.

Usage:
  dagger session [flags]

Flags:
  -h, --help           help for session
      --label labels   label that identifies the source of this session (e.g, --label 'sdk:python' --label 'sdk_version:0.5.2' --label 'sdk_async:true')

Global Flags:
      --debug            show buildkit debug logs
      --workdir string   The host workdir loaded into dagger (default ".")
```

### Follow-up
- [x] Let the `dagger session` command check if the run is happening from a CI or not (and append it `labels`)
```shell
./dagger session --label toto:tata
2023/03/28 16:22:33 deeebug:|toto:tata,ci:false,|
```

- [x] Let the `dagger session` command retrieve the anonymized (hashed) "git committer" and anonymized (hashed) "git repo"

### CODE REVIEW

**Review by gerhard — APPROVED** [review]
Looks good to me! WDYT @aluzzardi?

File: cmd/dagger/session.go [hunk 1]
```diff
@@ -2,52 +2,118 @@ package main
...
+	config, err := repo.Config()
+	if err != nil {
+		return "", "", err
+	}
+
+	committerEmail := config.User.Email
+	committerHash := fmt.Sprintf("%x", sha256.Sum256([]byte(committerEmail)))
+
+	remote, err := repo.Remote("origin")
+	var repoURL string
```
Comment (Chirab) [inline cmd/dagger/session.go:66]: I think we can remove this, wdyt ?

Comment (Chirab) [reply cmd/dagger/session.go:66]: I'm sorry if I expressed myself poorly.

I wanted to point `var repoURL string`

**Review by vito — COMMENTED** [review]
Looks good but I think the naming is a bit confusing.

We already have [pipeline labels](https://github.com/dagger/dagger/blob/1da83ba5c1af84222909723b850898dc18814514/core/pipeline/label.go), so I thought this PR was adding support for setting them on session start, but it looks like this is purely for tracking data in a user agent and doesn't actually propagate to pipeline labels.

I was extra confused to see we're collecting Git info here too, but it makes sense in hindsight, since here it's anonymized.

I would recommend just renaming everything to avoid the term "label" unless we're actually tying it into pipeline labels.

**Comment by gerhard** [comment]
>  We already have [pipeline labels](https://github.com/dagger/dagger/blob/1da83ba5c1af84222909723b850898dc18814514/core/pipeline/label.go), so I thought this PR was adding support for setting them on session start, but it looks like this is purely for tracking data in a user agent and doesn't actually propagate to pipeline labels.

That is actually a good idea! I can see us using a subset of these labels in the user agent, *and* submitting them all to Dagger Cloud when users connect their runs. Us being able to see which SDK was used for that pipeline run, version, etc. can be very useful for debugging purposes. That was one of the ideas behind the original Dagger Cloud implementation, and I still think that it's a valid one today. WDYT @shykes?

> I was extra confused to see we're collecting Git info here too, but it makes sense in hindsight, since here it's anonymized.

Yes, the fact that we are collecting the same info, just anonymising, is another hint that these should most likely be the same thing, just exposed differently based on where we are using it. There is no need to be doing the same thing in two different ways.

> I would recommend just renaming everything to avoid the term "label" unless we're actually tying it into pipeline labels.

Thanks for a great idea @vito. I'm going to try and make a few changes based on the above so that this is less confusing.

File: cmd/dagger/session.go [hunk 2]
```diff
@@ -98,3 +163,37 @@ func EngineSession(cmd *cobra.Command, args []string) error {
...
+func (ls *labels) Add(name string, value string) {
+	*ls = append(*ls, label{Name: name, Value: value})
+}
+
+func (ls *labels) String() string {
+	var labels string
+	for _, label := range *ls {
+		labels += fmt.Sprintf("%s:%s,", label.Name, label.Value)
+	}
+	return labels
```
[the remark this answers is not in the document]
Comment (gerhard) [reply cmd/dagger/session.go:198]: Yes, the comma is intentional. The end-result that we expect is `<LABELS>, go-containerregistry/v0.13.0`

**Comment by gerhard** [comment]
How does this look @vito? https://github.com/dagger/dagger/pull/4802/commits/043a08332007dee25d5ffb81ab72f7d78cd68382

**Review by vito — COMMENTED** [review]
Sorry, I didn't really communicate well before; I think I would prefer just calling all this "user agent" instead of "label" rather than unifying the two.

The two are definitely similar but I think aligning them too closely is more confusing than helpful. For example, we would likely never take the full set of `[]pipeline.Label` and pass it to the user agent string because it'd be way too big. And if `Engine` were to have a `Labels` field I would expect it to be a `[]pipeline.Label`, not a string encoded user agent fragment.

Injecting the user agent (or equivalent information) as a pipeline label on the other hand makes sense to me, as you said, so it feels like pipeline labels are a superset of user agent fragment. Maybe we could have the same code that initializes those pipeline labels just initialize the user agent fragment at the same time, and expose a getter for it like we have for `pipeline.RootLabels()`?

```go
&engine.Config{
	// ...
	UserAgent: pipeline.UserAgent(&sessionAgentFragments),
}
```

Under the hood `pipeline.UserAgent` could be synchronized with `pipeline.LoadRootLabels` (via the existing `*sync.Once`) so we only have to load git info and set up user agent fragment + labels once. (`LoadRootLabels` could probably be renamed at this point.)

What do you think?

**Comment by gerhard** [comment]
Yes, that makes sense @vito, and I'm all for improving this as a follow-up. For now, I would be inclined to have something which is good enough out, and come back to it next week.

`--user-agent` as a flag to the session doesn't make much sense, since that is an implementation detail specific to `ref.Get`. Surfacing it at the edge of the system - `dagger session` - seems leaky to me. From the outside, we are adding some metadata to the session. We can call it `--tag` or something similar, but I do see us wanting to push this info through all the way into the Cloud.

Would you prefer that we go with `dagger session --tag` instead of `dagger session --label`? Can you think of a better flag name?

**Pull request merged**